sys/conf/newvers.sh

```diff
@@ -263,6 +267,10 @@ if [ -n "$git_cmd" ] ; then
 	git=" ${git}"
 fi
 
+if [ -n "$gituprevision" ] ; then
+	gitup=" $(awk -F : '{printf "%s (%s)", $2, $1}' $gituprevision)"
```

Comment:
Can we mimic the git version a bit better here? 
That is, do stable/12-hash in case there is branch information. If it's a detached HEAD, then showing that is pretty useless and it should just fall back to the hash, in that case.

can you provide the full contents of a .gituprevision please? I'm curious to know what else is in there. Thanks!

Reply:
I don't consider `(detached)` a useless information. It is, in fact, a valueable one: it straightly says that the admin is not on a branch and will not receive updates will a pull. Moreover, if you need to bisect down to a bug it will tell and you have left some branch. This will clearly put the commit in context. I would even prefer to see this with the original `git(1)` output. GitHub shows you the same information: This commit does not belong to any branch.

I can live with `printf "%s-%s", $branch, $commit`. Though, I have intentionally not selected a hyphen because a colon cannot be part of a branch name and most shells break words on colons. See my reasoning [here](https://github.com/johnmehr/gitup/issues/19#issuecomment-753399723).

Reply:
An admin should use `git status` to see if they are detached or something. Furthermore, any `git bisect` command will cause a detached head anyway and sticking this in OSVERSION seems like the wrong way to go about this. We really only care what the srcdir contents were and whether they were modified from that. The branch name is already super redundant with the other version information in there, when dealing with "official" branches (like 12-STABLE). It's only there so that it can be clear that a user has their own branch, so any hash shown will likely not be in the official repo and thus someone trying to reproduce this will have a hard time.

Are you using `git status` regularly? Maybe as part of your shell's PS1 fanciness? I can highly recommend that.

I, for one, compile with something like this:

`cd /usr/src && git pull --rebase --autostash && git status && sleep 5 && nice make buildworld buildkernel`

Which prevents me from doing dumb things most of the time. The OSVERSION string needs to be short and to the point and we can't keep tacking on more on more stuff.

This machine here is still on an old svn build, and just look at that. Why do I need stable 12 in there 3 times, really? One FreeBSD would also likely suffice.

```
% uname -a
FreeBSD roadrunner.spoerlein.net 12.2-STABLE FreeBSD 12.2-STABLE r368787+6a45b9926f2e(stable/12) GENERIC  amd64
```

Reply:
The purpose of `gitup(1)` is not to use `git(1)` and making this for an admin as easy as possible. I personally don't see a need for an admin to learn Git if gitup just works.

> Are you using git status regularly? Maybe as part of your shell's PS1 fanciness? I can highly recommend that.

I am a developer, I use Git all the time and have a fancy `PS1` for that, but not talking about me. I am talking about non-programming admins.

> FreeBSD roadrunner.spoerlein.net 12.2-STABLE FreeBSD 12.2-STABLE r368787+6a45b9926f2e(stable/12) GENERIC  amd64

2x `12.2-STABLE` is purely a FreeBSD problem, not a SCM one. I am so see the revision as absolutely redundant because you have the commit anyway (this is not Subversion afterall). I still consider having the Git branch in the output as valuable in case of troubleshooting when people report bugs with FreeBSD.

Again, most of the admins aren't devs, don't expect them to wield it like you and me.

If you really care about no duplication, it should only contain the shortened commit hash that's it. The revision can always be calculcation from a commit with `git(1)`, but the branch cannot always be determined if it is detached.

Reply:
As I wrote, it's the old SVN+git version line, so don't read too much into it.

I still don't follow why the branch (or detached HEAD) is useful, if we have the hash, we know which branch that is on (or it's a purely local commit/branch and then all bets are off).

Reply:
@emaste What is your opinion?

Reply:
"Detached" is useless information. With the hash you know all you need to know. Git pull needs some thought at times, and this is one of them...  branch name is mildly useful to know how to interpret c1235..., but our c # will be small if you are on a local branch... I like to see it when I'm unsure which branch was on a ma chine. But absent bisects, I'm never running detached...

Reply:
@bsdimp OK, then lets drop the branch name, but I would expect that the branch name would be dropped also when `git(1)` is used for consistency reasons. Can we agree on that?

Reply:
I would agree. We also didn't have the branch-name in SVN. It can instead be derived from the 12-STABLE string, etc.

Reply:
> 
> 
> I would agree. We also didn't have the branch-name in SVN. It can instead be derived from the 12-STABLE string, etc.

Not always when the admin is using a custom branch which is not `stable/12`.

I will change the PR.

Reply:
@imp @uqs Done. Branch removed. We have the commit id only now.

Reply:
Shall I provide the change for `git(1)` also?

Reply:
Yes please, let's do it in 1 go. I think you can just remove the 4 lines around git_b

Can you then please provide the output of running newvers.sh in a gitup and a git checkout?

Reply:
This it how `gitup(1)` looks now:
```
$ uname -a
FreeBSD bsd1srv.fritz.box 12.2-STABLE FreeBSD 12.2-STABLE 43e9a678f GENERIC  amd64
```
with `git(1)`:
```
$ uname -a
FreeBSD bsd1srv.fritz.box 12.2-STABLE FreeBSD 12.2-STABLE c243246-g43e9a678f101 GENERIC  amd64
```
and here with a shallow clone:
```
$ uname -a
FreeBSD bsd1srv.fritz.box 12.2-STABLE FreeBSD 12.2-STABLE c1-g43e9a678f GENERIC  amd64
```

I consider this counter to be deceiving and not hard-reliable.

Pushed update.

Reply:
@uqs Please have a look now.